assignment4/count_islands.py

```diff
@@ -0,0 +1,94 @@
+class CountIslands():
+    def __init__(self):
+        self._clear_data()
+
+    def _clear_data(self):
+        """*_row_islands is a list of sets containing tile numbers that belong to some island in the previous row
+        """
+        self.prev_row_islands = []
+        self.new_row_islands = []
+        self.counter = 0
+        self._clear_iterators()
+
+    def _clear_iterators(self):
+        self.prev_row_i = 0
+        self.new_row_i = 0
+
+    def _create_island(self, tile_num):
+        """We either create new set (new island) or extend an existing one.
+        :param tile_num:
+        :return:
+        """
+        added = False
+        if not self.new_row_islands:
+            self.new_row_islands.append({tile_num})
+            return
+
+        for island in self.new_row_islands:
+            if tile_num - 1 in island:
+                island |= {tile_num}
+                added = True
+
+        if not added:
+            self.new_row_islands.append({tile_num})
+
+    def _compare_rows(self):
+        """We traverse two lists of sets to check how many sets in second list (self.new_row_islands)
+        do not intersect with any of the sets from list one (self.prev_row_islands).
+        We do it so that the complexity is O(n), incrementing one iterator or another, until we check
+        all the sets from list two.
+        :return:
+        """
+        if not self.prev_row_islands and self.new_row_islands:
+            self.counter += len(self.new_row_islands)
+            return
+        if not self.new_row_islands:
+            return
+
+        self._clear_iterators()
```

Comment:
Perhaps in this case there isn't any need to store the iterators as instance variables of the object, since you clear them every time anyway? Initializing variables inside the function (**new_row_i** and **prev_row_i**) seems to work equally well?